modules/margin-protocol/src/lib.rs

```diff
@@ -671,54 +670,86 @@ impl<T: Trait> Module<T> {
 	}
 
 	/// Margin level of a given user.
-	///
-	/// If `new_position` is `None`, return the margin level based on current positions, else based on current
-	/// positions plus this new one. If `equity_delta`, apply the delta to current equity.
-	fn _margin_level(
-		who: &T::AccountId,
-		new_position: Option<Position<T>>,
-		equity_delta: Option<Fixed128>,
-	) -> Fixed128Result {
-		let mut equity = Self::_equity_of_trader(who)?;
-		if let Some(d) = equity_delta {
-			equity = equity.checked_add(&d).ok_or(Error::<T>::NumOutOfBound)?;
+	fn _margin_level(who: &T::AccountId, action: Action<T>) -> Fixed128Result {
+		let equity = Self::_equity_of_trader(who)?;
+		let leveraged_debits_in_usd = Self::_leveraged_debits_in_usd(who)?;
+		// if no leveraged held, margin level is max
+		match action {
+			Action::Withdraw(amount) => {
+				let fixed128_amount = fixed_128_from_u128(amount);
+				// ensure there's no value loss
+				ensure!(
+					u128_from_fixed_128(fixed128_amount) == amount,
+					Error::<T>::NumOutOfBound
+				);
```

Comment:
Withdrawal overflowing is not going to happen, and if it happen somehow, saturating in `fixed_128_from_u128` would be enough.